### PR TITLE
SA2B: v2.3 - The Chao Update

### PR DESCRIPTION
**Pull request by PoryGone**
Changelog:

Features:
- New goal
  - Chaos Chao
    - Raise a Chaos Chao to win!
- New optional Location Checks
  - Chao Animal Parts
    - Each body part from each type of animal is a location
  - Chao Stats
    - 0-99 levels of each of the 7 Chao stats can be locations
    - The frequency of Chao Stat locations can be set (every level, every 2nd level, etc)
  - Kindergartensanity
    - Classroom lessons are locations
      - Either all lessons or any one of each category can be set as locations
  - Shopsanity
    - A specified number of locations can be placed in the Chao Black Market
    - These locations are unlocked by acquiring `Chao Coin`s
    - Ring costs for these items can be adjusted 
  - Chao Karate can now be set to one location per fight, instead of one per tournament
- Items
  - If any Chao locations are active, the following will be in the item pool:
    - Chao Eggs
    - Garden Seeds
    - Garden Fruit
    - Chao Hats
    - Chaos Drives
- The starting eggs in the garden can be a random color
- Chao World entrances can be shuffled
- Chao are given default names
- New Traps
  - Reverse Trap

Quality of Life:
- Chao Save Data is now separate per-slot in addition to per-seed
  - This allows a single player to have multiple slots in the same seed, each having separate Chao progress
- Chao Race/Karate progress is now displayed on Stage Select (when hovering over Chao World)
- All Chao can now enter the Hero and Dark races
- Chao Karate difficulty can be set separately from Chao Race difficulty
- Chao Aging can be sped up at will, up to 15×
- New mod `config` option to fine-tune Chao Stat multiplication
  - Note: This does not mix well with the Mod Manager "`Chao Stat Multiplier`" code
- Pong Traps can now activate in Chao World
- Maximum range for possible number of Emblems is now 1000
- General APWorld cleanup and optimization
  - Option access has moved to the new options system
  - An item group now exists for trap items

Bug Fixes:
- Dry Lagoon now has all 11 Animals
- Eternal Engine - 2 (Standard and Hard Logic) now requires only `Tails - Booster`
- Lost Colony - 2 (Hard Logic) now requires no upgrades
- Lost Colony - Animal 9 (Hard Logic) now requires either `Eggman - Jet Engine` or `Eggman - Large Cannon`

### CODE REVIEW

File: worlds/sa2b/Items.py [hunk 1]
```diff
@@ -107,12 +245,24 @@ def __init__(self, name, classification: ItemClassification, code: int = None, p
...
+    **animals_table,
+    **chaos_drives_table,
     **event_table,
 }
 
 lookup_id_to_name: typing.Dict[int, str] = {data.code: item_name for item_name, data in item_table.items() if data.code}
 
-item_groups: typing.Dict[str, str] = {"Chaos Emeralds": [item_name for item_name, data in emeralds_table.items()]}
+item_groups: typing.Dict[str, str] = {
+    "Chaos Emeralds": [item_name for item_name, data in emeralds_table.items()],
```
Comment (Berserker66) [inline worlds/sa2b/Items.py:260]: gut  feeling says list(emeralds_table), or list(emeralds_table.keys()) if you want to be explicit should be faster.

File: worlds/sa2b/Missions.py [hunk 2]
```diff
@@ -193,10 +194,10 @@
     "Cannon's Core - ",
 ]
 
-def get_mission_count_table(multiworld: MultiWorld, player: int):
+def get_mission_count_table(multiworld: MultiWorld, world: World, player: int):
     mission_count_table: typing.Dict[int, int] = {}
 
-    if multiworld.goal[player] == 3:
+    if world.options.goal == 3:
```
Comment (Berserker66) [inline worlds/sa2b/Missions.py:200]:
```suggestion
    if world.options.goal == "grand_prix":
```

Are you aware you can write this? It's a bit easier to follow when one does not know that 3 happens to be grand prix.

**Review by Berserker66 — APPROVED** [review]
code looks good to me, left some optional hints for improvement.

File: worlds/sa2b/Missions.py (hunk 2 above)
Comment (alwaysintreble) [reply worlds/sa2b/Missions.py:200]: Alternatively, could `from .options import Goal` and then do `world.options.goal == Goal.option_grand_prix` if you want the typing which is how I prefer it.

**Review by Berserker66 — APPROVED** [review]

**Pull request merged**